### PR TITLE
error propagation for API calls

### PR DESCRIPTION
**Pull request by harshdeeppruthi**

### CODE REVIEW

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowAPIException.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package io.cdap.plugin.servicenow.apiclient;
+
+import org.apache.http.HttpResponse;
+
+import javax.annotation.Nullable;
+
+/**
+ * Javadoc.
+ */
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowAPIException.java]: Please add proper documentation

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 2]
```diff
@@ -112,22 +128,32 @@ private static void validateRestApiResponse(String responseBody) {
...
     if (SUCCESS_CODES.contains(code)) {
-      return;
+      return null;
     }
     if (RETRYABLE_CODES.contains(code)) {
-      throw new RetryableException(String.format(HTTP_ERROR_MESSAGE, code));
+      return new ServiceNowAPIException(String.format(HTTP_ERROR_MESSAGE, code), response, true);
     }
-    throw new NonRetryableException(String.format(HTTP_ERROR_MESSAGE, code));
+    return new ServiceNowAPIException(String.format(HTTP_ERROR_MESSAGE, code), response, false);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: You can directly use
return new ServiceNowAPIException(String.format(HTTP_ERROR_MESSAGE, code), response, RETRYABLE_CODES.contains(code))

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java [hunk 3]
```diff
@@ -112,22 +128,32 @@ private static void validateRestApiResponse(String responseBody) {
     }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: We should remove the RetryableException and NonRetyableException classes completely and use the ServiceNowApiException object directly as used below.

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowAPIException.java (hunk 1 above)
Comment (harshdeeppruthi) [reply src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowAPIException.java]: Done.

File: src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java (hunk 3 above)
Comment (harshdeeppruthi) [reply src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java]: Done.

File: src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowAPIException.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    this.manualRetry = false;
+
+  }
+
+  public ServiceNowAPIException(String message, Throwable t,
+      @Nullable HttpResponse httpResponse, boolean manualRetry) {
+    super(message, t);
+    this.httpResponse = httpResponse;
+    this.manualRetry = manualRetry;
+  }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/servicenow/apiclient/ServiceNowAPIException.java]:
```suggestion
  public ServiceNowAPIException(String message, @Nullable HttpResponse httpResponse) {
    this(message, null, httpResponse, false);
  }

  public ServiceNowAPIException(Throwable t, @Nullable HttpResponse httpResponse) {
    this(null, t, httpResponse, false)
  }

  public ServiceNowAPIException(String message, Throwable t,
      @Nullable HttpResponse httpResponse, boolean manualRetry) {
    super(message, t);
    this.httpResponse = httpResponse;
    this.manualRetry = manualRetry;
  }
```

**Review by MrRahulSharma — APPROVED** [review]

**Pull request merged**